fbpic/particles/gathering/threading_methods.py

```diff
@@ -382,220 +383,237 @@ def gather_field_numba_cubic(x, y, z,
-                else:
-                    Fz_m += Sz[index_z]*Sr[index_r]* \
-                        Ez_m0[iz[index_z], ir[index_r]]
-                index_z += 1
-            index_r += 1
-
-        Fr += (Fr_m*exptheta_m0).real
-        Ft += (Ft_m*exptheta_m0).real
-        Fz += (Fz_m*exptheta_m0).real
-
-        # Mode 1
-        # ----------------------------
-        # Clear the temporary variables
-        # for the "per mode" gathering
-        Fr_m = 0.j
-        Ft_m = 0.j
-        Fz_m = 0.j
-        # Add the fields for mode 1
-        index_r = 0
-        while index_r < 4:
-            index_z = 0
-            while index_z < 4:
-                if Sz[index_z]*Sr[index_r] < 0:
-                    Fr_m += (-1.)*Sz[index_z]*Sr[index_r]* \
-                                Er_m1[iz[index_z], ir[index_r]]
-                    Ft_m += (-1.)*Sz[index_z]*Sr[index_r]* \
-                                Et_m1[iz[index_z], ir[index_r]]
-                else:
-                    Fr_m += Sz[index_z]*Sr[index_r]* \
-                                Er_m1[iz[index_z], ir[index_r]]
-                    Ft_m += Sz[index_z]*Sr[index_r]* \
-                                Et_m1[iz[index_z], ir[index_r]]
-                Fz_m += Sz[index_z]*Sr[index_r]*Ez_m1[iz[index_z], ir[index_r]]
-                index_z += 1
-            index_r += 1
-
-        # Add the fields from the mode 1
-        Fr += 2*(Fr_m*exptheta_m1).real
-        Ft += 2*(Ft_m*exptheta_m1).real
-        Fz += 2*(Fz_m*exptheta_m1).real
 
-        # Convert to Cartesian coordinates
-        # and write to particle field arrays
-        Ex[i] = (cos*Fr - sin*Ft)
-        Ey[i] = (sin*Fr + cos*Ft)
-        Ez[i] = Fz
-
-        # B-Field
-        # ----------------------------
-        # Clear the placeholders for the
-        # gathered field for each coordinate
-        Fr = 0.
-        Ft = 0.
-        Fz = 0.
+    nthreads : int
+        Number of CPU threads used with numba prange
 
-        # Mode 0
-        # ----------------------------
-        # Create temporary variables
-        # for the "per mode" gathering
-        Fr_m = 0.j
-        Ft_m = 0.j
-        Fz_m = 0.j
-        # Add the fields for mode 0
-        index_r = 0
-        while index_r < 4:
+    ptcl_chunk_indices : array of int, of size nthreads+1
+        The indices (of the particle array) between which each thread
+        should loop. (i.e. divisions of particle array between threads)
+    """
+    # Gather the field per cell in parallel
+    for nt in prange( nthreads ):
+
+        # Create private arrays for each thread
+        # to store the particle index and shape
+        ir = np.empty( 4, dtype=int64)
+        Sr = np.empty( 4 )
+        iz = np.empty( 4, dtype=int64)
+        Sz = np.empty( 4 )
+
+        # Loop over all particles in thread chunk
+        for i in range( ptcl_chunk_indices[nt],
+                            ptcl_chunk_indices[nt+1] ):
+
```

Comment:
The only change to this function is the above modification to the loop.  (Looping over nthreads and then over particles within each chunk, instead of looping directly over particles.)

The rest of this function is unchanged: it is simply indented to the right because of the new nested loop. (the only exception is the `np.zeros` calls which were removed from the body of the function below)